multiregionccl: introduce data-driven test framework for multi-region

arulajmani opened the pull request:
This patch introduces a new data-driven test framework for multi-region
syntax. `TestMultiRegionDataDriven` contains a full description of how
to use this thing.

There were a few motivations to introduce this. Firstly, this makes it
very easy to create specific cluster topologies, which is quite useful
when testing multi-region invariants at a high level. Secondly, this
allows us to create directives that analyze and assert invariants in
query traces easily. This also allows us to create directives to wait
for zone config updates to take place which is quite neat.

I've also added a test for global tables that uses this framework. The
test ensures that once new regions are added to the database, present
time reads can be served via (local) follower reads from the new
region.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/65313)
<!-- Reviewable:end -->

arulajmani committed:
multiregionccl: introduce data-driven test framework for multi-region

This patch introduces a new data-driven test framework for multi-region
syntax. TestMultiRegionDataDriven contains a full description of how
to use this thing.

There were a few motivations to introduce this. Firstly, this makes it
very easy to create specific cluster topologies, which is quite useful
when testing multi-region invariants at a high level. Secondly, this
allows us to create directives that analyze and assert invariants in
query traces easily. This also allows us to create directives to wait
for zone config updates to take place which is quite neat.

I've also added a test for global tables that uses this framework. The
test ensures that once new regions are added to the database, present
time reads can be served via (local) follower reads from the new
region.

Release note: None

arulajmani reviewed (commented):
Yo @otan, this is now in a much better place and should be ready for a look. The key here was introducing this "wait-for-zone-config-changes" directive, which makes the whole thing deterministic. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65313#-:-M_x9ay19BwVogFgeLPh:b-ws609r)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)



<!-- Sent from Reviewable.io -->

otan reviewed (commented):
Looking good, curious about the broken test though.
Some changes to fatal desired.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65313#-:-M_xf35K2D3FmxABqJHF:b-mbn5c9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[pkg/ccl/multiregionccl/datadriven_test.go, line 48 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65313#-M_xdy2k-qeQv2WZZneq:-M_xdy2k-qeQv2WZZner:b-gddtrn) ([raw file](https://github.com/cockroachdb/cockroach/blob/208e69c83cbb06435cf02b19859cf8dac7c952a9/pkg/ccl/multiregionccl/datadriven_test.go#L48)):*
> ```Go
> // provided may later be used to index into a particular node to run a query.
> //
> // "set-server idx=server_number": set the node at server_number (in the order
> ```

can you mention whether it is 0 or 1 indexed
___
*[pkg/ccl/multiregionccl/datadriven_test.go, line 104 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65313#-M_xewKQ7bCxnNgA9fT3:-M_xewKQ7bCxnNgA9fT4:by2j704) ([raw file](https://github.com/cockroachdb/cockroach/blob/208e69c83cbb06435cf02b19859cf8dac7c952a9/pkg/ccl/multiregionccl/datadriven_test.go#L104)):*
> ```Go
> 			case "new-cluster":
> 				if ds.tc != nil {
> 					return errors.New("cluster already exists, cleanup cluster first").Error()
> ```

this should fatal
___
*[pkg/ccl/multiregionccl/datadriven_test.go, line 111 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65313#-M_xexh2CCmzqzlEDOzB:-M_xexh2CCmzqzlEDOzC:by2j704) ([raw file](https://github.com/cockroachdb/cockroach/blob/208e69c83cbb06435cf02b19859cf8dac7c952a9/pkg/ccl/multiregionccl/datadriven_test.go#L111)):*
> ```Go
> 				}
> 				if localities == "" {
> 					return errors.Newf("can't create a cluster without %s", serverLocalities).Error()
> ```

this should fatal
___
*[pkg/ccl/multiregionccl/datadriven_test.go, line 163 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65313#-M_xeuna-ieitYx_zk8T:-M_xeuna-ieitYx_zk8U:by2j704) ([raw file](https://github.com/cockroachdb/cockroach/blob/208e69c83cbb06435cf02b19859cf8dac7c952a9/pkg/ccl/multiregionccl/datadriven_test.go#L163)):*
> ```Go
> 			case "query-sql":
> 				if d.HasArg(expectFollowerRead) && !d.HasArg(expectServeLocally) {
> 					return errors.Newf(
> ```

this should fatal
___
*[pkg/ccl/multiregionccl/datadriven_test.go, line 242 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65313#-M_xefG8B-lCCKcdub_x:-M_xefG8B-lCCKcdub_y:by2j704) ([raw file](https://github.com/cockroachdb/cockroach/blob/208e69c83cbb06435cf02b19859cf8dac7c952a9/pkg/ccl/multiregionccl/datadriven_test.go#L242)):*
> ```Go
> 			case "refresh-range-descriptor-cache":
> 				if !d.HasArg(tableName) {
> 					return errors.Newf("no %q provided", tableName).Error()
> ```

this should fatal
___
*[pkg/ccl/multiregionccl/datadriven_test.go, line 245 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65313#-M_xefdRCCzzZje39slh:-M_xefdRCCzzZje39sli:by2j704) ([raw file](https://github.com/cockroachdb/cockroach/blob/208e69c83cbb06435cf02b19859cf8dac7c952a9/pkg/ccl/multiregionccl/datadriven_test.go#L245)):*
> ```Go
> 				}
> 				if !d.HasArg(serverIdx) {
> 					return errors.Newf("no %s provided", serverIdx).Error()
> ```

this should fatal
___
*[pkg/ccl/multiregionccl/datadriven_test.go, line 263 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65313#-M_xeYOO570eRwG4X-92:-M_xeYOO570eRwG4X-93:b-5qyker) ([raw file](https://github.com/cockroachdb/cockroach/blob/208e69c83cbb06435cf02b19859cf8dac7c952a9/pkg/ccl/multiregionccl/datadriven_test.go#L263)):*
> ```Go
> 				// Ensure the range descriptor cache was indeed populated.
> 				var tableID uint32
> 				err = ds.curDB.QueryRow(fmt.Sprintf(`SELECT id from system.namespace WHERE name='%s'`, tbName)).Scan(&tableID)
> ```

nit: should we use `$1` here to do proper escaping? below too.
___
*[pkg/ccl/multiregionccl/datadriven_test.go, line 277 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65313#-M_xed_yBC0W16cpKorN:-M_xed_yBC0W16cpKorO:by2j704) ([raw file](https://github.com/cockroachdb/cockroach/blob/208e69c83cbb06435cf02b19859cf8dac7c952a9/pkg/ccl/multiregionccl/datadriven_test.go#L277)):*
> ```Go
> 			case "wait-for-zone-config-changes":
> 				if !d.HasArg(tableName) {
> 					return errors.Newf("no %s provided", tableName).Error()
> ```

this should fatal
___
*[pkg/ccl/multiregionccl/datadriven_test.go, line 307 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65313#-M_xej-q6yaTABDwJlSY:-M_xej-q6yaTABDwJlSZ:b-hbrjpy) ([raw file](https://github.com/cockroachdb/cockroach/blob/208e69c83cbb06435cf02b19859cf8dac7c952a9/pkg/ccl/multiregionccl/datadriven_test.go#L307)):*
> ```Go
> 					}
> 					if !found {
> 						errors.Newf("did not find leaseholder for %s", tbName)
> ```

return?
___
*[pkg/ccl/multiregionccl/datadriven_test.go, line 390 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65313#-M_xeKqcDnkA-UCly3Ut:-M_xeKqcDnkA-UCly3Uu:biofuk8) ([raw file](https://github.com/cockroachdb/cockroach/blob/208e69c83cbb06435cf02b19859cf8dac7c952a9/pkg/ccl/multiregionccl/datadriven_test.go#L390)):*
> ```Go
> 	"us-east-1": {
> 		Tiers: []roachpb.Tier{
> 			{Key: "region", Value: "us-east-1"},
> ```

can we put some AZs in here?


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/ccl/multiregionccl/datadriven_test.go, line 307 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65313#-M_xej-q6yaTABDwJlSY:-M_z1c56E6WW5SveIe0f:b-nfo1m) ([raw file](https://github.com/cockroachdb/cockroach/blob/208e69c83cbb06435cf02b19859cf8dac7c952a9/pkg/ccl/multiregionccl/datadriven_test.go#L307)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

return?
</blockquote></details>

(wonder if it's the source of your flake)


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
Okay, more digging, and heres what I have. This was flaking because of 2 reasons:

1. When we ran these things through the replicate queue, if the leaseholder was not one of the nodes we expect to have a voting replica, replicas weren't as expected. To fix this, I've to transfer the lease to a node that we expect to be a voting replica and run it through the queue again.
 
2. There's a legitimate issue with how we optimize replica order such that we may skip the local replica (thereby making a network hop when serving a follower read even if we don't need to). There was some chatter about this on slack a couple weeks back, and the fix is easy, so I'll do that in a separate patch.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65313#-:-Ma-eFCG-Fvbxqb8W_h6:b-35no28)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/ccl/multiregionccl/datadriven_test.go, line 48 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65313#-M_xdy2k-qeQv2WZZneq:-Ma-MeNi8TJ75xVuZv8r:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/208e69c83cbb06435cf02b19859cf8dac7c952a9/pkg/ccl/multiregionccl/datadriven_test.go#L48)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

can you mention whether it is 0 or 1 indexed
</blockquote></details>

Done.
___
*[pkg/ccl/multiregionccl/datadriven_test.go, line 104 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65313#-M_xewKQ7bCxnNgA9fT3:-Ma-MpKx5aoScA2FvMrA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/208e69c83cbb06435cf02b19859cf8dac7c952a9/pkg/ccl/multiregionccl/datadriven_test.go#L104)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

this should fatal
</blockquote></details>

Done.
___
*[pkg/ccl/multiregionccl/datadriven_test.go, line 111 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65313#-M_xexh2CCmzqzlEDOzB:-Ma-Mt9U-ZHRMz4fTPkp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/208e69c83cbb06435cf02b19859cf8dac7c952a9/pkg/ccl/multiregionccl/datadriven_test.go#L111)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

this should fatal
</blockquote></details>

Done.
___
*[pkg/ccl/multiregionccl/datadriven_test.go, line 163 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65313#-M_xeuna-ieitYx_zk8T:-Ma-My6aCJUfDzOWn-3-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/208e69c83cbb06435cf02b19859cf8dac7c952a9/pkg/ccl/multiregionccl/datadriven_test.go#L163)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

this should fatal
</blockquote></details>

Done.
___
*[pkg/ccl/multiregionccl/datadriven_test.go, line 242 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65313#-M_xefG8B-lCCKcdub_x:-Ma-N3ls8TCRSJHUiQ39:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/208e69c83cbb06435cf02b19859cf8dac7c952a9/pkg/ccl/multiregionccl/datadriven_test.go#L242)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

this should fatal
</blockquote></details>

Done.
___
*[pkg/ccl/multiregionccl/datadriven_test.go, line 245 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65313#-M_xefdRCCzzZje39slh:-Ma-N4D-5gD769K11pjC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/208e69c83cbb06435cf02b19859cf8dac7c952a9/pkg/ccl/multiregionccl/datadriven_test.go#L245)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

this should fatal
</blockquote></details>

Done.
___
*[pkg/ccl/multiregionccl/datadriven_test.go, line 277 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65313#-M_xed_yBC0W16cpKorN:-Ma-NKVCB_z2xY5nr2Yn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/208e69c83cbb06435cf02b19859cf8dac7c952a9/pkg/ccl/multiregionccl/datadriven_test.go#L277)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

this should fatal
</blockquote></details>

Done.
___
*[pkg/ccl/multiregionccl/datadriven_test.go, line 307 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65313#-M_xej-q6yaTABDwJlSY:-Ma-eAPS1gC-N2M7OARu:b-y083fw) ([raw file](https://github.com/cockroachdb/cockroach/blob/208e69c83cbb06435cf02b19859cf8dac7c952a9/pkg/ccl/multiregionccl/datadriven_test.go#L307)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

(wonder if it's the source of your flake)
</blockquote></details>

Thanks for catching. But na, not the source of the flake haha
___
*[pkg/ccl/multiregionccl/datadriven_test.go, line 390 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65313#-M_xeKqcDnkA-UCly3Ut:-Ma-dBPp9DrrGD3OfkTI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/208e69c83cbb06435cf02b19859cf8dac7c952a9/pkg/ccl/multiregionccl/datadriven_test.go#L390)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

can we put some AZs in here?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

arulajmani commented:
Looks like @andreimatei  already has a PR out for this https://github.com/cockroachdb/cockroach/pull/64394/files

nvb reviewed (commented):
> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65313)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @otan)

___
*[pkg/ccl/multiregionccl/datadriven_test.go, line 56 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/65313#-MaZSxdz4As7KsdhtqxA:-MaZSxe-6mIxeTTk4cNQ:b-s9t9za) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ffb53f44a5480770fc17facbf027dde0c1c70f0/pkg/ccl/multiregionccl/datadriven_test.go#L56)):*
> ```Go
> // using "set-server".
> //
> // "query-sql [idx=server_number] [serve-locally] [serve-follower-reads]:
> ```

I'm curious about the decision to make `serve-locally` and `serve-follower-reads` directives, instead of having them be part of the command output. The most useful part of datadriven tests is that they make it possible to use the `--rewrite` flag to rewrite expected behavior without any manual intervention. I generally tend to use this as a guiding question to determine how to structure inputs and outputs. The question to ask is "how do I structure outputs so that behavior changes are captured by a --rewrite?". This leads to fewer assertion-style tests, and more trace-style tests. Examples of cases where this has gone well are in the optimizer tests, which print out explain plans. An example where this has gone less well is in logictests, where error messages are part of the command directives.

So you can imagine that a variant of what we have here that would be more amenable to `--rewrite`-ing would be:
```
query-sql idx=4
SELECT * FROM db.global WHERE k = 1
----
served locally: true
served via follower read: true
```
This would allow us to bulk rewrite a suite of queries as we improve locality-optimized search and easily demonstrate improvements. It would also allow us to extend the attributes we assert against without needing to touch each test case in the future. For instance, if we wanted to distinguish between intra-region queries and intra-zone queries, we easily could (e.g. `served from same zone: false`). Or if we wanted to count cross-region RTTs to demonstrate improvements from a change like https://github.com/cockroachdb/cockroach/issues/64872, we could (e.g. `cross-region network requests: 3`).
___
*[pkg/ccl/multiregionccl/datadriven_test.go, line 61 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/65313#-MaZSaht9NzIV_x764N4:-MaZSaht9NzIV_x764N5:b-ud3f2t) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ffb53f44a5480770fc17facbf027dde0c1c70f0/pkg/ccl/multiregionccl/datadriven_test.go#L61)):*
> ```Go
> //    - serve-locally: ensures that the query was routed to the local
> //    replica.
> //    -	serve-follower-reads: ensures that the query was served using a follower
> ```

nit: strange spacing here
___
*[pkg/ccl/multiregionccl/datadriven_test.go, line 62 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/65313#-MaZSe44Ah8KqvDK-lCJ:-MaZSe44Ah8KqvDK-lCK:b8vfofx) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ffb53f44a5480770fc17facbf027dde0c1c70f0/pkg/ccl/multiregionccl/datadriven_test.go#L62)):*
> ```Go
> //    replica.
> //    -	serve-follower-reads: ensures that the query was served using a follower
> //    read. Currently, this flag only works when "serve-locally-soon" is provided
> ```

`serve-locally-soon` no longer exists
___
*[pkg/ccl/multiregionccl/datadriven_test.go, line 148 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/65313#-MaZWMN4-z07heOvbESb:-MaZWMN52IWmeopgFN6P:b-gnhd2x) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ffb53f44a5480770fc17facbf027dde0c1c70f0/pkg/ccl/multiregionccl/datadriven_test.go#L148)):*
> ```Go
> 
> 			case "cleanup-cluster":
> 				ds.cleanup(ctx)
> ```

nit: new-line after this.
___
*[pkg/ccl/multiregionccl/datadriven_test.go, line 151 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/65313#-MaZWVIZ61PERWtIB236:-MaZWVIZ61PERWtIB237:bnetfa5) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ffb53f44a5480770fc17facbf027dde0c1c70f0/pkg/ccl/multiregionccl/datadriven_test.go#L151)):*
> ```Go
> 			case "exec-sql":
> 				sqlDB := ds.curDB
> 				if d.HasArg(serverIdx) {
> ```

Is it worth supporting cases where the serverIdx is not provided? Or will that cause more confusion than its worth?
___
*[pkg/ccl/multiregionccl/datadriven_test.go, line 369 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/65313#-MaZWz-R6MpGodXFPD1T:-MaZWz-R6MpGodXFPD1U:bcp9vca) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ffb53f44a5480770fc17facbf027dde0c1c70f0/pkg/ccl/multiregionccl/datadriven_test.go#L369)):*
> ```Go
> 				}
> 
> 			case "set-server":
> ```

We don't currently use this. Do we plan to?
___
*[pkg/ccl/multiregionccl/datadriven_test.go, line 378 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/65313#-MaZWrquELQSEuD9wPoQ:-MaZWrquELQSEuD9wPoR:b-gnhd2x) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ffb53f44a5480770fc17facbf027dde0c1c70f0/pkg/ccl/multiregionccl/datadriven_test.go#L378)):*
> ```Go
> 					return err.Error()
> 				}
> 				ds.curDB = sqlDB
> ```

nit: new-line after this.


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65313)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten and @otan)

___
*[pkg/ccl/multiregionccl/datadriven_test.go, line 56 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/65313#-MaZSxdz4As7KsdhtqxA:-Mb2t9ZRBStKIZMT1P3E:bpikcoh) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ffb53f44a5480770fc17facbf027dde0c1c70f0/pkg/ccl/multiregionccl/datadriven_test.go#L56)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I'm curious about the decision to make `serve-locally` and `serve-follower-reads` directives, instead of having them be part of the command output. The most useful part of datadriven tests is that they make it possible to use the `--rewrite` flag to rewrite expected behavior without any manual intervention. I generally tend to use this as a guiding question to determine how to structure inputs and outputs. The question to ask is "how do I structure outputs so that behavior changes are captured by a --rewrite?". This leads to fewer assertion-style tests, and more trace-style tests. Examples of cases where this has gone well are in the optimizer tests, which print out explain plans. An example where this has gone less well is in logictests, where error messages are part of the command directives.

So you can imagine that a variant of what we have here that would be more amenable to `--rewrite`-ing would be:
```
query-sql idx=4
SELECT * FROM db.global WHERE k = 1
----
served locally: true
served via follower read: true
```
This would allow us to bulk rewrite a suite of queries as we improve locality-optimized search and easily demonstrate improvements. It would also allow us to extend the attributes we assert against without needing to touch each test case in the future. For instance, if we wanted to distinguish between intra-region queries and intra-zone queries, we easily could (e.g. `served from same zone: false`). Or if we wanted to count cross-region RTTs to demonstrate improvements from a change like https://github.com/cockroachdb/cockroach/issues/64872, we could (e.g. `cross-region network requests: 3`).
</blockquote></details>

I initially had the statement trace analysis inside of a succeeds soon, which is why I had to accept these as directives. I was calling it "serve-locally-soon" then which is why you saw some remnants of it in the comment below. I just never revisited that decision once I'd stabilized this test to the point that I didn't need a succeeds soon. 

I really like your suggestion, I hadn't really considered the power of `--rewrite` here. Changed the test so that it now looks similar to the illustration above. 
___
*[pkg/ccl/multiregionccl/datadriven_test.go, line 151 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/65313#-MaZWVIZ61PERWtIB236:-Mb2qykT7C6f7xOxcneh:b64gyfl) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ffb53f44a5480770fc17facbf027dde0c1c70f0/pkg/ccl/multiregionccl/datadriven_test.go#L151)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Is it worth supporting cases where the serverIdx is not provided? Or will that cause more confusion than its worth?
</blockquote></details>

Removed, I'm happy with tests having to explicitly specify the server index. 
___
*[pkg/ccl/multiregionccl/datadriven_test.go, line 369 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/65313#-MaZWz-R6MpGodXFPD1T:-Mb2qarR07cmFKP3MEfa:b11jjk0) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ffb53f44a5480770fc17facbf027dde0c1c70f0/pkg/ccl/multiregionccl/datadriven_test.go#L369)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We don't currently use this. Do we plan to?
</blockquote></details>

I agree with your previous comment, there's not much value in supporting a set-server pattern. Getting rid of this entirely, now all commands expect an index argument explicitly. 


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65313#-:-MbCHd8m4rJ27df_HvCN:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan)



<!-- Sent from Reviewable.io -->

arulajmani commented:
TFTR! 

bors r=nvanbenschoten
